server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```diff
@@ -217,24 +217,32 @@ private void doRecovery(final long recoveryId, final StartRecoveryRequest preExi
         final RecoveryTarget recoveryTarget = recoveryRef.target();
         assert recoveryTarget.sourceNode() != null : "cannot do a recovery without a source node";
         final RecoveryState.Timer timer = recoveryTarget.state().getTimer();
+        final IndexShard indexShard = recoveryTarget.indexShard();
+        final boolean promotableToPrimary = indexShard.routingEntry().isPromotableToPrimary();
```

Comment:
I think I'd rather have a completely separate (and much simpler) path for `promotableToPrimary == false` rather than adding conditionals deep within the existing flow like this.

Reply:
I tried doing that but it ended up having a lot of copied code. I believe it's only two `ifs` below. But if you think it's important, would you be able to give a suggestion of how to do it?

Reply:
Something like this perhaps?

```java
    private void doRecovery(final long recoveryId, final StartRecoveryRequest preExistingRequest) {
        final RecoveryRef recoveryRef = onGoingRecoveries.getRecovery(recoveryId);
        if (recoveryRef == null) {
            logger.trace("not running recovery with id [{}] - can not find it (probably finished)", recoveryId);
            return;
        }
        final RecoveryTarget recoveryTarget = recoveryRef.target();
        assert recoveryTarget.sourceNode() != null : "cannot do a recovery without a source node";
        final RecoveryState.Timer timer = recoveryTarget.state().getTimer();
        final RecoveryState recoveryState = recoveryTarget.state();
        final IndexShard indexShard = recoveryTarget.indexShard();

        final var failureHandler = ActionListener.notifyOnce(ActionListener.runBefore(
            ActionListener.noop().delegateResponse((l, e) -> {
                // this will be logged as warning later on...
                logger.trace("unexpected error while preparing shard for peer recovery, failing recovery", e);
                onGoingRecoveries.failRecovery(
                    recoveryId,
                    new RecoveryFailedException(recoveryTarget.state(), "failed to prepare shard for recovery", e),
                    true
                );
            }), recoveryRef::close));

        if (indexShard.routingEntry().isPromotableToPrimary() == false) {
            assert preExistingRequest == null;
            assert indexShard.indexSettings().getIndexMetadata().isSearchableSnapshot() == false;
            try {
                indexShard.preRecovery(failureHandler.delegateFailure((l, v) -> ActionListener.completeWith(l, () -> {
                    logger.trace("{} preparing shard for peer recovery", recoveryTarget.shardId());
                    indexShard.prepareForIndexRecovery();
                    // Skip unnecessary intermediate stages
                    recoveryState.setStage(RecoveryState.Stage.VERIFY_INDEX);
                    recoveryState.setStage(RecoveryState.Stage.TRANSLOG);
                    indexShard.openEngineAndSkipTranslogRecovery();
                    recoveryState.getIndex().setFileDetailsComplete();
                    recoveryState.setStage(RecoveryState.Stage.FINALIZE);
                    onGoingRecoveries.markRecoveryAsDone(recoveryId);
                    return null;
                })));
            } catch (Exception e) {
                failureHandler.onFailure(e);
            }

            return;
        }

        record StartRecoveryRequestToSend(StartRecoveryRequest startRecoveryRequest, String actionName, TransportRequest requestToSend) {}
        final ActionListener<StartRecoveryRequestToSend> toSendListener = failureHandler.map(r -> {
            logger.trace(
                "{} [{}]: recovery from {}",
                r.startRecoveryRequest().shardId(),
                r.actionName(),
                r.startRecoveryRequest().sourceNode()
            );
            transportService.sendRequest(
                r.startRecoveryRequest().sourceNode(),
                r.actionName(),
                r.requestToSend(),
                new RecoveryResponseHandler(r.startRecoveryRequest(), timer)
            );
            return null;
        });

        if (preExistingRequest == null) {
            try {
                indexShard.preRecovery(toSendListener.delegateFailure((l, v) -> ActionListener.completeWith(l, () -> {
                    logger.trace("{} preparing shard for peer recovery", recoveryTarget.shardId());
                    indexShard.prepareForIndexRecovery();
                    if (indexShard.indexSettings().getIndexMetadata().isSearchableSnapshot()) {
                        // for searchable snapshots, peer recovery is treated similarly to recovery from snapshot
                        indexShard.getIndexEventListener().afterFilesRestoredFromRepository(indexShard);
                        final Store store = indexShard.store();
                        store.incRef();
                        try {
                            StoreRecovery.bootstrap(indexShard, store);
                        } finally {
                            store.decRef();
                        }
                    }
                    final long startingSeqNo = indexShard.recoverLocallyUpToGlobalCheckpoint();
                    assert startingSeqNo == UNASSIGNED_SEQ_NO || recoveryTarget.state().getStage() == RecoveryState.Stage.TRANSLOG
                        : "unexpected recovery stage [" + recoveryTarget.state().getStage() + "] starting seqno [ " + startingSeqNo + "]";
                    final var startRequest = getStartRecoveryRequest(logger, clusterService.localNode(), recoveryTarget, startingSeqNo);
                    return new StartRecoveryRequestToSend(startRequest, PeerRecoverySourceService.Actions.START_RECOVERY, startRequest);
                })));
            } catch (Exception e) {
                toSendListener.onFailure(e);
            }
        } else {
            toSendListener.onResponse(
                new StartRecoveryRequestToSend(
                    preExistingRequest,
                    PeerRecoverySourceService.Actions.REESTABLISH_RECOVERY,
                    new ReestablishRecoveryRequest(recoveryId, preExistingRequest.shardId(), preExistingRequest.targetAllocationId())
                )
            );
        }
    }
```

Reply:
Ohh nice! Ah, the experience with the ActionListeners shines :100: 

I'll adopt this, thanks.